### ISSUE TITLE
Devcontainer not working on M1 Mac

### ISSUE THREAD

**Issue by creecedev** [hd-zero/hdzero-goggle#64]
When starting the remote dev container, the setup.sh runs and after the media is extracted, there's a CMake error that a simple test program cannot be compiled. The same process of cloning into the container and building on my Windows machine had no issues.

I am new to Docker in general but i'm doing the same thing on both systems and getting different results.

C++/CMake/CMake Tools are installed as extensions within the dev container.

[variant] Loaded new set of variants
[proc] Executing command: /usr/bin/aarch64-linux-gnu-gcc -v
[proc] Executing command: /usr/bin/aarch64-linux-gnu-gcc-11 -v
[proc] Executing command: /usr/bin/c89-gcc -v
[proc] Executing command: /usr/bin/c99-gcc -v
[proc] Executing command: /usr/bin/gcc -v
[proc] Executing command: /usr/bin/gcc-11 -v
[proc] Executing command: /bin/aarch64-linux-gnu-gcc -v
[proc] Executing command: /bin/aarch64-linux-gnu-gcc-11 -v
[proc] Executing command: /bin/c89-gcc -v
[proc] Executing command: /bin/c99-gcc -v
[proc] Executing command: /bin/gcc -v
[proc] Executing command: /bin/gcc-11 -v
[proc] Executing command: /bin/c99-gcc -v
[proc] Executing command: /bin/c89-gcc -v
[proc] Executing command: /usr/bin/c99-gcc -v
[proc] Executing command: /usr/bin/c89-gcc -v
[kit] Found Kit: GCC 11.3.0 aarch64-linux-gnu
[kit] Found Kit: GCC 11.3.0 aarch64-linux-gnu
[kit] Found Kit: GCC 11.3.0 aarch64-linux-gnu
[kit] Found Kit: GCC 11.3.0 aarch64-linux-gnu
[kit] Found Kit: GCC 11.3.0 aarch64-linux-gnu
[kit] Found Kit: GCC 11.3.0 aarch64-linux-gnu
[kit] Found Kit: GCC 11.3.0 aarch64-linux-gnu
[kit] Found Kit: GCC 11.3.0 aarch64-linux-gnu
[kit] Found Kit: GCC 11.3.0 aarch64-linux-gnu
[kit] Found Kit: GCC 11.3.0 aarch64-linux-gnu
[kit] Found Kit: GCC 11.3.0 aarch64-linux-gnu
[kit] Found Kit: GCC 11.3.0 aarch64-linux-gnu
[proc] Executing command: /usr/bin/aarch64-linux-gnu-gcc -v
[cmakefileapi-driver] Removing /workspaces/hdzero-goggle/build/CMakeCache.txt
[main] Configuring folder: hdzero-goggle 
[proc] Executing command: /usr/bin/cmake --no-warn-unused-cli -DCMAKE_TOOLCHAIN_FILE=/workspaces/hdzero-goggle/toolchain.cmake -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE -DCMAKE_BUILD_TYPE:STRING=Debug -DCMAKE_C_COMPILER:FILEPATH=/usr/bin/aarch64-linux-gnu-gcc -DCMAKE_CXX_COMPILER:FILEPATH=/usr/bin/aarch64-linux-gnu-g++ -S/workspaces/hdzero-goggle -B/workspaces/hdzero-goggle/build -G Ninja
[cmake] Not searching for unused variables given on the command line.
[cmake] -- The C compiler identification is unknown
[cmake] -- The CXX compiler identification is unknown
[cmake] -- Detecting C compiler ABI info
[cmake] -- Detecting C compiler ABI info - failed
[cmake] -- Check for working C compiler: /workspaces/hdzero-goggle/toolchain/bin/arm-openwrt-linux-muslgnueabi-gcc
[cmake] -- Check for working C compiler: /workspaces/hdzero-goggle/toolchain/bin/arm-openwrt-linux-muslgnueabi-gcc - broken
[cmake] -- Configuring incomplete, errors occurred!
[cmake] See also "/workspaces/hdzero-goggle/build/CMakeFiles/CMakeOutput.log".
[cmake] See also "/workspaces/hdzero-goggle/build/CMakeFiles/CMakeError.log".
[cmake] CMake Error at /usr/share/cmake-3.22/Modules/CMakeTestCCompiler.cmake:69 (message):
[cmake]   The C compiler
[cmake] 
[cmake]     "/workspaces/hdzero-goggle/toolchain/bin/arm-openwrt-linux-muslgnueabi-gcc"
[cmake] 
[cmake]   is not able to compile a simple test program.
[cmake] 
[cmake]   It fails with the following output:
[cmake] 
[cmake]     Change Dir: /workspaces/hdzero-goggle/build/CMakeFiles/CMakeTmp
[cmake]     
[cmake]     Run Build Command(s):/usr/bin/ninja cmTC_d1209 && [1/2] Building C object CMakeFiles/cmTC_d1209.dir/testCCompiler.c.o
[cmake]     FAILED: CMakeFiles/cmTC_d1209.dir/testCCompiler.c.o 
[cmake]     /workspaces/hdzero-goggle/toolchain/bin/arm-openwrt-linux-muslgnueabi-gcc   -Wall -o CMakeFiles/cmTC_d1209.dir/testCCompiler.c.o -c /workspaces/hdzero-goggle/build/CMakeFiles/CMakeTmp/testCCompiler.c
[cmake]     qemu-x86_64: Could not open '/lib64/ld-linux-x86-64.so.2': No such file or directory
[cmake]     ninja: build stopped: subcommand failed.
[cmake]     
[cmake]     
[cmake] 
[cmake]   
[cmake] 
[cmake]   CMake will not be able to correctly generate this project.
[cmake] Call Stack (most recent call first):
[cmake]   CMakeLists.txt:3 (project)
[cmake] 
[cmake] 
[proc] The command: /usr/bin/cmake --no-warn-unused-cli -DCMAKE_TOOLCHAIN_FILE=/workspaces/hdzero-goggle/toolchain.cmake -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE -DCMAKE_BUILD_TYPE:STRING=Debug -DCMAKE_C_COMPILER:FILEPATH=/usr/bin/aarch64-linux-gnu-gcc -DCMAKE_CXX_COMPILER:FILEPATH=/usr/bin/aarch64-linux-gnu-g++ -S/workspaces/hdzero-goggle -B/workspaces/hdzero-goggle/build -G Ninja exited with code: 1 and signal: null

**Comment by creecedev** [comment]
This appears to be related to M1 Macbooks. I found a stackoverflow that discusses the issue:

https://stackoverflow.com/questions/71040681/qemu-x86-64-could-not-open-lib64-ld-linux-x86-64-so-2-no-such-file-or-direc

The solution was to add --platform=linux/amd64 to the FROM line of the Dockerfile:

**FROM --platform=linux/amd64 mcr.microsoft.com/devcontainers/base:ubuntu**

I'm now able to successfully compile on my M1 Macbook. I'm not sure if this breaks anything if added for other systems but it may be worth adding if not.

**Comment by bkleiner** [comment]
appears to me that linking itself succeeds, but something fails during packaging.
`qemu: uncaught target signal 11 (Segmentation fault) - core dumped` seems the place where it fails.
i dont see any usage of qemu in the build script. is that part of the M1 emulation somehow?

**Comment by creecedev** [comment]
> 

Yes, it appears its something to do with M1. I'm just going to go back to Windows and develop there as I have no idea how to get this one resolved.

**Comment by bkleiner** [comment]
turns out a lot of the problematic process was redundant. could you try https://github.com/hd-zero/hdzero-goggle/pull/111 and see if it now builds on M1.

**Comment by Nikolas-S** [comment]
> turns out a lot of the problematic process was redundant. could you try #111 and see if it now builds on M1?

Not sure when it was fixed but works fine for me in an x86 container on M1, it's just really slow since it's emulated.
I tried getting an arm native toolchain to work and actually managed to get the build working in an arm64 virtualized container by using the arm gnu toolchain and kinda smashing it together with the x86 one. 
It builds a lot faster but I'd be surprised if it actually runs, if anyone feels like trying it, I take no responsibility. ^^
https://github.com/Nikolas-S/hdzero-goggle/tree/feature/arm64_host_build_support
